executor/executor.c

```diff
@@ -0,0 +1,491 @@
+*  the Python function call. If loop is nonzero, this will block the calling thread forever.
+*  This function should be run in a separate thread.
+* 
+*  Inputs: Uses thread_args_t as input struct.
+*      Necessary fields:
+*          func_name: string of function name to run in the student code
+*          mode: string of the current mode
+*          loop: boolean for whether the Python function should be called in a while loop forever
+*      Optional fields:
+*          cond: condition variable that blocks the calling thread, used with `run_function_timeout`
+*          timeout: max length of execution time before a warning is issued for the function call
+*/
+void* run_py_function(void *thread_args) {
+	thread_args_t *args = (thread_args_t *)thread_args;
+
+	PyGILState_STATE gstate = PyGILState_Ensure();
+	log_runtime(DEBUG, "Acquired GIL");
+	pthread_cleanup_push(gil_cleanup, (void*) &gstate);
+	
+	//assign the run mode to the Gamepad object
+	PyObject *pMode = PyUnicode_FromString(args->mode);
+	int err = PyObject_SetAttrString(pGamepad, "mode", pMode);
+	Py_DECREF(pMode);
+	if (err != 0) {
+		PyErr_Print();
+		char msg[128];
+		sprintf(msg, "Couldn't assign mode for Gamepad while trying to run %s", args->func_name);
+		log_runtime(ERROR, msg);
+		if (args->cond != NULL) {
+			pthread_cond_signal(args->cond);
+		}
+		return NULL;
+	}
+	
+	//retrieve the Python function from the student code
+	PyObject *pFunc = PyObject_GetAttrString(pModule, args->func_name);
+	PyObject *pValue = NULL;
+	if (pFunc && PyCallable_Check(pFunc)) {
+		struct timespec start, end;
+		long time, max_time;
+		if (args->timeout) {
+			max_time = args->timeout->tv_sec * 1e9 + args->timeout->tv_nsec;
+		}
+		//Necessary cleanup to handle pthread_cancel 
+		struct cleanup objects = { pFunc, pValue };
+		pthread_cleanup_push(py_function_cleanup, (void *) &objects);
+		pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
+		
+		do {
+			clock_gettime(CLOCK_MONOTONIC, &start);
+			pValue = PyObject_CallObject(pFunc, NULL); //make call to Python function
+			clock_gettime(CLOCK_MONOTONIC, &end);
+
+			//if the time the Python function took was greater than max_time, warn that it's taking too long
+			time = (end.tv_sec - start.tv_sec) * 1e9 + (end.tv_nsec - start.tv_nsec);
+			if (args->loop && args->timeout && time > max_time) {
+				char msg[128];
+				sprintf(msg, "Function %s is taking longer than %lu milliseconds, indicating a loop in the code.", args->func_name, (long) (max_time / 1e6));
+				log_runtime(WARN, msg);
+			}
+			//catch execution error
+			if (pValue == NULL) {
+				if (PyErr_Occurred()) {
+					PyErr_Print();
+				}
+				char msg[128];
+				sprintf(msg, "Python function %s call failed", args->func_name);
+				log_runtime(ERROR, msg);
+				break;
+			}
+			else {
+				Py_DECREF(pValue);
+			}
+		} while (args->loop);
+		pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, NULL);

```

Comment:
Is there a reason why we wait until here to set the cancel type back to `DEFERRED`? I thought we only want it asynchronous when the Python function is called, in `PyObject_CallObject()`... if that's the case, the two calls to `pthread_canceltype()` should probably go between lines 259 and 260, and between lines 260 and 261, to wrap the call to the Python function.

Reply:
Fixed